src/EnergyPlus/Data/EnergyPlusData.hh

```diff
@@ -0,0 +1,184 @@
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#ifndef EnergyPlusData_hh_INCLUDED
+#define EnergyPlusData_hh_INCLUDED
+
+// EnergyPlus Headers
+#include <EnergyPlus/EnergyPlus.hh>
+
+#include <string>
+
+struct BaseGlobalStruct
+{
+    virtual void clear_state() = 0;
+};
+
+//struct OutputReportTabular : BaseGlobalStruct
+//{
+//    //int MaxHeaderLength;
+//
+//    void clear_state() override {
+//    }
+//};
+
+struct DataGlobal : BaseGlobalStruct
+{
+    // Data
+    bool AnnualSimulation = false;
+
+    // MODULE VARIABLE DECLARATIONS:
+    std::string DayOfSimChr = "0";       // Counter for days (during the simulation) (character -- for reporting)
+
+    // MODULE PARAMETER DEFINITIONS
+    static constexpr int EndZoneSizingCalc = 4;
+
+    void clear_state() override {
+        AnnualSimulation = false;
+        DayOfSimChr = "0";
+    }
+};
+
+struct ExteriorEnergyUseData : BaseGlobalStruct
+{
+    struct ExteriorLightUsage
+    {
+        // Members
+        std::string Name;          // Descriptive name -- will show on reporting
+        int SchedPtr;              // Can be scheduled
+        Real64 DesignLevel;        // Consumption in Watts
+        Real64 Power;              // Power = DesignLevel * ScheduleValue
+        Real64 CurrentUse;         // Use for this time step
+        int ControlMode;           // Control mode Schedule Only or Astronomical Clock plus schedule
+        bool ManageDemand;         // Flag to indicate whether to use demand limiting
+        Real64 DemandLimit;        // Demand limit set by demand manager [W]
+        bool PowerActuatorOn;      // EMS flag
+        Real64 PowerActuatorValue; // EMS value
+        Real64 SumConsumption;     // sum of electric consumption [J] for reporting
+        Real64 SumTimeNotZeroCons; // sum of time of positive electric consumption [hr]
+
+                                   // Default Constructor
+        ExteriorLightUsage()
+            : SchedPtr(0), DesignLevel(0.0), Power(0.0), CurrentUse(0.0), ControlMode(1), ManageDemand(false), DemandLimit(0.0),
+            PowerActuatorOn(false), SumConsumption(0.0), SumTimeNotZeroCons(0.0)
+        {
+        }
+    };
+
+    int NumExteriorLights = 0; // Number of Exterior Light Inputs
+    int NumExteriorEqs = 0;    // Number of Exterior Equipment Inputs
+
+    // Object Data
+    Array1D<ExteriorLightUsage> ExteriorLights;        // Structure for Exterior Light reporting
+
+    void clear_state() override {
+        NumExteriorLights = 0;
+        NumExteriorEqs = 0;
+        ExteriorLights.deallocate();

```

Comment:
What happens if `deallocate` is never called? Is that a memory leak or does `Array1D` do the right thing when the ExteriorLights are destroyed?

Reply:
I'm not sure I can answer this properly, but I'll try.

When EnergyPlus runs normally, the clear_state function is never called.  The program ends and all the resources are presumably all destroyed and the program is done.

The reasons you would want to call clear_state are:
1. You are running the unit test suite and you want to clear the state before running another unit test, or
2. You are using EnergyPlus as a library and you want to clear the state between simulations.

In either case, you want to reset the state back to the initialized state, which would primarily include zeroing variables and calling deallocate on everything on the heap, so that the next run will be able to allocate it again.

Based on those use cases, can you give me a bit more info on your concern about deallocate never being called?

Reply:
Building on my previous comment about ```clear_state```s. Here the ```clear_state``` is implemented but is missing most of the members in the data class. It would be great if we could automatically enforce that all variables get added here. I'm unsure how that would happen.

Reply:
I'm not sure I'm seeing that @mitchute.  The struct is there with members, but the only variables that hold state are the two integers and the array of lights, and those three are all reset here.  I could be wrong, let me know if you see something I'm missing.  Also, yes, having a way to ensure everything is cleared would be great, but even if we can't do that automatically, just having the code organized in these compact state structs should make it much easier to identify variables that aren't reset.